src/client/csfle/client_encryption.rs

```diff
@@ -250,25 +330,123 @@ pub struct ClientEncryptionOptions {
+        mut self,
+        provider: KmsProvider,
+        config: bson::Document,
+    ) -> ClientEncryptionBuilder<BUILDER_READY> {
+        self.opts.kms_providers.insert(provider, config);
+        ClientEncryptionBuilder { opts: self.opts }
+    }
+
+    /// Add configuration for multiple KMS providers.  Already-added configuration for other
+    /// providers will be preserved.
+    pub fn extend_kms_providers(
+        mut self,
+        providers: impl IntoIterator<Item = (KmsProvider, bson::Document)>,
+    ) -> ClientEncryptionBuilder<BUILDER_READY> {
+        self.opts.kms_providers.extend(providers);
+        ClientEncryptionBuilder { opts: self.opts }
+    }
+
+    /// Add TLS options for a single KMS provider.
+    pub fn tls_options(mut self, provider: KmsProvider, options: TlsOptions) -> Self {
+        self.opts
+            .tls_options
+            .get_or_insert_with(KmsProvidersTlsOptions::new)
+            .insert(provider, options);
+        self
+    }
+
+    /// Add TLS options for multiple KMS providers.  Already-added configuration for other providers
+    /// will be preserved.
+    pub fn extend_tls_options(
+        mut self,
+        options: impl IntoIterator<Item = (KmsProvider, TlsOptions)>,
+    ) -> Self {
+        self.opts
+            .tls_options
+            .get_or_insert_with(KmsProvidersTlsOptions::new)
+            .extend(options);
+        self
+    }
+}
+
+impl ClientEncryptionBuilder<BUILDER_READY> {
+    /// Create a new key vault handle with the given options.
+    pub fn build(self) -> Result<ClientEncryption> {
+        ClientEncryption::new(self.opts)
+    }
+}
+
 /// Options for creating a data key.
-#[derive(Debug, Clone, TypedBuilder)]
-#[builder(field_defaults(setter(into)))]
+#[derive(Debug, Clone)]
 #[non_exhaustive]
-pub struct DataKeyOptions {
-    /// The master key document, a KMS-specific key used to encrypt the new data key.
-    pub master_key: MasterKey,
-    /// An optional list of alternate names used to reference a key.
-    #[builder(default)]
-    pub key_alt_names: Option<Vec<String>>,
-    /// An optional buffer of 96 bytes to use as custom key material for the data key being
+pub(crate) struct DataKeyOptions {
+    pub(crate) master_key: MasterKey,
+    pub(crate) key_alt_names: Option<Vec<String>>,
+    pub(crate) key_material: Option<Vec<u8>>,
+}
+
+/// A pending `ClientEncryption::create_data_key` action.
+pub struct CreateDataKeyAction<'a> {
+    client_enc: &'a ClientEncryption,
+    opts: DataKeyOptions,
+}
+
+impl<'a> CreateDataKeyAction<'a> {
+    /// Execute the pending data key creation.
+    pub async fn run(self) -> Result<Binary> {
+        self.client_enc
+            .create_data_key_final(&self.opts.master_key.provider(), self.opts)
+            .await
+    }
+
+    /// Set an optional list of alternate names that can be used to reference the key.
+    pub fn key_alt_names(mut self, names: impl IntoIterator<Item = String>) -> Self {
```

Comment:
Would we want to do the `key_alt_name` + `extend_key_alt_names` approach here too?